### PR TITLE
Autoloader: Fix Uninstallation Fatal

### LINKED ISSUE #18996: Autoloader: Exception when uninstalling Jetpack
#### Steps to reproduce the issue

By UI
1. Go to Admin -> Plugins in Wordpress
2. Install Jetpack
3. Uninstall Jetpack
4. See error in php error logs

By WP-CLI
1. Install Jetpack `wp plugin install jetpack --activate`
2. Uninstall Jetpack `wp plugin delete jetpack`
3. See error in console

#### What I expected
No errors, just uninstall the plugin

#### What happened instead
The plugin is uninstalled, but the error occurs

```
Fatal error: Uncaught RuntimeException: Failed to locate plugin C:\Users\guyasyou\Desktop\wordpress\wp-content\plugins\jetpack in C:\Users\guyasyou\Desktop\wordpress\wp-content\plugins\jetpack\vendor\jetpack-autoloader\class-plugin-locator.php:48
Stack trace:
#0 C:\Users\guyasyou\Desktop\wordpress\wp-content\plugins\jetpack\vendor\jetpack-autoloader\class-plugins-handler.php(99): Automattic\Jetpack\Autoloader\jpf11009ded9fc4592b6a05b61ce272b3cjetpack\Plugin_Locator->find_current_plugin()
#1 C:\Users\guyasyou\Desktop\wordpress\wp-content\plugins\jetpack\vendor\jetpack-autoloader\class-autoloader.php(106): Automattic\Jetpack\Autoloader\jpf11009ded9fc4592b6a05b61ce272b3cjetpack\Plugins_Handler->get_active_plugins(false, true)
#2 C:\Users\guyasyou\Desktop\wordpress\wp-includes\class-wp-hook.php(287): Automattic\Jetpack\Autoloader\jpf11009ded9fc4592b6a05b61ce272b3cjetpack\Autoloader::Automattic\Jetpack\Autoloader\jpf11009ded9fc4592b6a05b61ce272b3cjetpack\{closure}('')
#3 C:\Users\guyasyou\Desktop\wordpress\wp-includes\class-wp in C:\Users\guyasyou\Desktop\wordpress\wp-content\plugins\jetpack\vendor\jetpack-autoloader\class-plugin-locator.php on line 48
PS C:\Users\guyasyou\Desktop\wordpress>
```

The error occurs because of a shutdown hook that is created before uninstall and run after uninstall.
https://github.com/Automattic/jetpack/blob/master/projects/packages/autoloader/src/class-autoloader.php#L80

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
Fixes #18996

#### Changes proposed in this Pull Request:

When `Plugin_Locator::find_current_plugin()` is unable to find the plugin containing the autoloader it throws an exception. Normally this should never happen but during uninstallation a plugin might be deleted in the same request it is deactivated. When this happens it won't be able to locate the plugin with the autoloader and will throw the exception. This PR takes the safest approach and clears the autoloader cache so that the next request can rebuild it (if necessary).

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Just noting here that I've added a test case covering the bug.

* Install Jetpack `wp plugin install jetpack --activate`
* Uninstall Jetpack `wp plugin delete jetpack`
* Verify no error is present in console.

#### Proposed changelog entry for your changes:
* Autoloader: Fix uninstallation fatal.

### CODE REVIEW

File: projects/packages/autoloader/tests/php/tests/unit/ShutdownHandlerTest.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			->with( false, true )
+			->willReturn( array( TEST_PLUGIN_DIR ) );
+		$this->plugins_handler->expects( $this->once() )
+			->method( 'cache_plugins' )
+			->with( array( TEST_PLUGIN_DIR ) );
+
+		// Mark that the plugins have been loaded so that we can perform a safe shutdown.
+		do_action( 'plugins_loaded' );
+
+		( new Shutdown_Handler( $this->plugins_handler, array(), false ) )();
```
Comment (jeherve) [inline projects/packages/autoloader/tests/php/tests/unit/ShutdownHandlerTest.php:51]: Looks like you'll need to solve this so the tests don't fail.

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/packages/autoloader/tests/php/tests/unit/ShutdownHandlerTest.php (hunk 1 above)
Comment (kbrown9) [reply projects/packages/autoloader/tests/php/tests/unit/ShutdownHandlerTest.php:51]: It looks like that syntax isn't compatible with PHP 5.6. We can probably just assign the object to a variable first then invoke it.

Comment (ObliviousHarmony) [reply projects/packages/autoloader/tests/php/tests/unit/ShutdownHandlerTest.php]: I've fixed the test.

**Review by kbrown9 — APPROVED** [review]
The changes look good to me. I tested by reproducing the error on the master branch and confirming that the error is not generated with this branch.

**Pull request merged**